### DISCUSSION TITLE
How to run junit test suite in Tycho

### DISCUSSION THREAD

**Discussion by Monstereye** [eclipse-tycho/tycho#953] [General]
Hello

To be hones I don't know if this is the right place to ask my question. If not the just tell me and/or delete the discussion.

I have a rather huge eclipse rcp project which recently migrated to a maven tycho based build. All works fine except of our unit tests.
The tests ar, up to now, organised in a way that each plugin has its own test fragment.
Within each test fragment is a test suite that "collects" all the tests.
Above all is a "main" test plugin that has a test suite consisting of all the other test suites (which then containing the tests itselfs).

My problem is now that I don't understand how to run a setup like this?
What is now the best concept to let the unit tests run?
Maven surefire itself is not working because dependency resolution has to be done via manifest.mf and tycho.
Using tycho-surerfire is also not possible/"needed" because I don't need an OSGI runtime for the tests (so far at least) and if I try to let it run as such the OSGI runtime is crashing with missing dependencies.
Is this maybe an option? : [https://github.com/eclipse/tycho/blob/master/RELEASE_NOTES.md#execute-unit-tests-with-eclipse-plugin-packaging](https://github.com/eclipse/tycho/blob/master/RELEASE_NOTES.md#execute-unit-tests-with-eclipse-plugin-packaging)

If so then I do something wrong:
I add the fragments as a module inside of pom.
Then I did exactly (hopefully correct :-) ) what is described there.
Problem is that the tests are not running. There is no surefire report.

Am I do something wrong?
What is the best concept to have unit tests in fragments?

If the concept with the test suites is not good for maven approach then this is no problem at all.
This is quickly removed.

To be honest I thought having tests in a separate fragment is rather a standard for eclipse rcp applications? (If not please correct me :-))
I was surprised to not find any examples how to setup such an approach.
Therefore my questions here. 

I added my pom as text file (and removed the modules because I wanted to concentrate on the configuration.
It's a text file because uploading a xml file was restricted.

[pom.txt](https://github.com/eclipse/tycho/files/8654726/pom.txt)

If there is any information missing you might want to have to be able to answer my question please tell me.

Hoping for an answer and many many thanks in advance

Sebastian Heinrich

**Comment by laeubi** [comment]
I see no obvious reason why it should not work, but fear you better provide a small reproducer project for more information or help what could be wrong.

**Comment by phermsdorf** [comment]
a simple setup with test execution can be found here: https://github.com/eclipse/tycho/tree/master/demo/itp01
the tests are in a bundle not in a fragment but basically it should work the same way

**Comment by Monstereye** [comment]
Many thanks for your comments and I am sorry for my very late answer. 
I take care about the project not in fulltime so please forgive me that I need sometimes more time to answer.

Back to topic:
I will have a closer look into the suggested example and will create a small test project to demonstrate.
This will take a little bit of time so please be patient with me :-)

**Comment by Monstereye** [comment]
First of all: I was not able to solve my problem although I think I made some progress.

Repeat of what I want:
Tests located in fragments to plugins, dependency resolution via Tycho and just one call to a single TestSuice calling all the tests (several thousands of unit tests).

The first problem was that as soon tycho-surefire-plugin detected a test fragment it startet to execute the tests contained.
This lead then to an error because most of the tests in all test fragments are somewhat related to database communication and the persistence.xml and orm.xml and so on are configured in the main test plugin (therefore the test suite to not repeat database configuration in each fragment).
Here I found with <skipTests>true</skipTests> a solution.

Now the test fragments are compiling! First step done.

The problem is now that my main test plugin fails to compile because of missing dependencies.
I tried to add the fragments to the target platform with 
   `<plugin>
		    <groupId>org.eclipse.tycho</groupId>
		    <artifactId>target-platform-configuration</artifactId>
			<version>${tycho-version}</version>
		    <configuration>
		        <dependency-resolution>
		            <extraRequirements>
		               <requirement>
		                    <type>eclipse-plugin</type>` 	

but as soon as I add the fragments here, the build is complaining that the host plugins are missing.
I pushed an minimum example here: 
[https://github.com/Monstereye/tycho-test-example.git](https://github.com/Monstereye/tycho-test-example.git)

This approach I found e.g. here [https://stackoverflow.com/questions/45190063/add-fragment-to-tycho-surefire-plugin#45190630](https://stackoverflow.com/questions/45190063/add-fragment-to-tycho-surefire-plugin#45190630) and here 
[https://bugs.eclipse.org/bugs/show_bug.cgi?id=436617#c11](https://bugs.eclipse.org/bugs/show_bug.cgi?id=436617#c11)
According to the hints in the mentioned links I added the fragments AND the host plugins as suggested. This only leads then to the error (from my minimum example) :
[DEBUG] No solution found because the problem is unsatisfiable.: [Unable to satisfy dependency from tycho.demo.itp.application.test 1.0.0.202205291007 to osgi.bundle; tycho.demo.itp.application 1.0.0.qualifier.; Unable to satisfy dependency from tycho.demo.itp02.test 1.0.0.202205291007 to osgi.bundle; tycho.demo.itp02 1.0.0.qualifier.; No solution found because the problem is unsatisfiable.]
[INFO] {osgi.os=linux, osgi.ws=gtk, org.eclipse.update.install.features=true, osgi.arch=x86_64}
[ERROR] Cannot resolve project dependencies:
[ERROR]   Software being installed: tycho.demo.itp.application.test 1.0.0.202205291007
[ERROR]   Missing requirement: tycho.demo.itp.application.test 1.0.0.202205291007 requires 'osgi.bundle; tycho.demo.itp.application 1.0.0.qualifier' but it could not be found
[ERROR] 
[ERROR] See https://wiki.eclipse.org/Tycho/Dependency_Resolution_Troubleshooting for help.
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for parent 1.0.0-SNAPSHOT:
[INFO] 
[INFO] parent ............................................. SUCCESS [  0.144 s]
[INFO] tycho.demo.itp.application ......................... SUCCESS [  3.203 s]
[INFO] tycho.demo.itp.application.test .................... SUCCESS [  1.060 s]
[INFO] tycho.demo.itp02 ................................... SUCCESS [  0.481 s]
[INFO] tycho.demo.itp02.test .............................. SUCCESS [  1.520 s]
[INFO] tycho.demo.itp.tests ............................... FAILURE [  1.935 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------

But in general:

I have plugins with fragments. The unit tests are contained within the fragments.
I would expect tycho to resolve these dependencies by itself. Should be possible, since the host plugin is mentioned in MANIFEST.MF?!
Otherwise, in my real project, I would have to add 200 Plugins + 200 test fragments and all their dependencies to the pom.xml of the main test plugin. This sound like a uge effort to maintain for me.

In the meantime I stumbled across this stackOverflow entry:
[https://stackoverflow.com/questions/13454654/how-to-get-a-fragment-bundle-into-tycho-test-runtime](https://stackoverflow.com/questions/13454654/how-to-get-a-fragment-bundle-into-tycho-test-runtime)
Therefore I removed the host plugin minimum version (which whas 1.0.0.qualifier) in all fragments.

In addition I followed this one (the section on eclipse-test-plugin packaging)
[https://wiki.eclipse.org/Tycho/Packaging_Types#eclipse-test-plugin](https://wiki.eclipse.org/Tycho/Packaging_Types#eclipse-test-plugin)
and the example here [https://github.com/gamerson/liferay-ide-sandbox/tree/master/tycho-test-runtime](https://github.com/gamerson/liferay-ide-sandbox/tree/master/tycho-test-runtime) 
and created a feature that contains all test fragments and their host plugins (in my minimum example).
And tada : it compiles and the tests run accordingly in tycho-surefire.

I will now tests this on my huge project and we will see.
The minimum example is adapted accordingly and will maybe help somebody in the future.

Two questions are remaining:

1. Is the feature really necessary? Why is tycho not able to resolve the dependencies of the test fragments automaticaally? In my case the main test suites has class entries for the unit tests located in the fragments. Therefore the host plugins of these fragments are contained in Manifest.mf dependencies list.
2. Why is tycho-surefire not automatically resolving the .qualifier of the fragments host plugin minimum version? Or at least throwing an appropriate warning/error message to do so. The mentioned stackOverflow entry from above is from 2013... 

I hope this will help somebody in the future and avoids reading and searching as much as I needed to do. :-)

**Comment by Monstereye** [comment]
I stumble over another problem.
As soon as I try to add a dependency (in Manifest.mf) to mockito in the test plugins the tycho complains about to not be able to resolve the dependency.

I tried several approaches only one of them worked.

Let me explain what I tried so far (all is included in the project here: https://github.com/Monstereye/tycho-test-example.git )
First I added the dependencies to the target file (from eclipse orbit). This target is set as active target platform and eclipse p2 build has no problem to resolve the dependencies.

Second I tried the following
1. This [[described](https://www.vogella.com/tutorials/EclipseTycho/article.html)](https://www.vogella.com/tutorials/EclipseTycho/article.html) says to add test dependencies in tycho-surefire plugin. I did as described.
Tycho still complains about not finding mockito.
2. This stackOverflow entry states to add such dependencies to target platform via a feature. [https://stackoverflow.com/questions/25476595/tycho-cannot-resolve-dependency-configured-in-tycho-surefire-plugin](https://stackoverflow.com/questions/25476595/tycho-cannot-resolve-dependency-configured-in-tycho-surefire-plugin)
3. I tried to add mockito as extra dependency in master pom - still not working.
4. Then I tried to resolve mockito via maven central and add it to the target file. Then, and only then, the tycho build is able to resolve mockito.

The problem with the approach in 4. is that the target file is a targlet generated via oomph.
The oomph setup file is (up to now) not able to have entries directly from maven central since (as far as I understood) it is only working with p2 repositorys.

Is there somebody how may have some time to look at my simple test project and find out why tycho-surefire is not able to resolve mockito?
Please note: Within the repo there are all approaches at once in the pom xmls. Therefore to reproduce the error:
Please remove the maven entries from the target file contained in  tycho.demo.itp.target

Any help is highly appreciated!
